### PR TITLE
Fixed #32779 -- Added the db_tablespace parameter to class-based unique constraints.

### PR DESCRIPTION
**Pull request by hannseman**
Ticket: https://code.djangoproject.com/ticket/32779

### CODE REVIEW

File: django/db/backends/base/schema.py [hunk 1]
```diff
@@ -1246,14 +1250,15 @@ def create_unique_name(*args, **kwargs):
...
-        if condition or include or opclasses or expressions:
+        if condition or include or opclasses or expressions or db_tablespace:
             sql = self.sql_create_unique_index
         else:
             sql = self.sql_create_unique
         if columns:
             columns = self._index_columns(table, columns, col_suffixes=(), opclasses=opclasses)
         else:
             columns = Expressions(table, expressions, compiler, self.quote_value)
+        tablespace_sql = self._get_index_tablespace_sql(model, fields=[], db_tablespace=db_tablespace)
```
Comment (hannseman) [inline django/db/backends/base/schema.py:1261]: I passed `fields` an empty list here as we don't have access to the actual field instances but instead. Felt that this might be OK since the auto-mapping of tablespaces fields and unique constraints never've been in place. If anyone thinks otherwise it looks like we could drop the `columns` argument to `_create_unique_sql` and add `fields` with actual field instances but not really sure how that would affect third-party backends.

File: django/db/models/base.py [hunk 2]
```diff
@@ -2063,6 +2063,24 @@ def _check_constraints(cls, databases):
...
+                    checks.Warning(
+                        '%s does not support setting a custom tablespace.' % connection.display_name,
+                        hint=(
+                            "A constraint won't be created. Silence this "
+                            "warning if you don't care about it."
+                        ),
+                        obj=cls,
+                        id='models.W045',
+                    )
+                )
```
Comment (felixxm) [inline django/db/models/base.py:2083]: IMO this check is not necessary, we should simply ignore `db_tablespace` when a database doesn't support it.

File: django/db/backends/base/schema.py [hunk 3]
```diff
@@ -1275,11 +1281,11 @@ def _delete_unique_sql(
             ) or
             (condition and not self.connection.features.supports_partial_indexes) or
             (include and not self.connection.features.supports_covering_indexes) or
-            (expressions and not self.connection.features.supports_expression_indexes)
-
+            (expressions and not self.connection.features.supports_expression_indexes) or
+            (db_tablespace and not self.connection.features.supports_tablespaces)
```
Comment (felixxm) [inline django/db/backends/base/schema.py:1285]: I would revert this change and ignore the `db_tablespace` option when a database doesn't support it.

File: docs/ref/models/constraints.txt [hunk 4]
```diff
@@ -203,3 +203,14 @@ For example::
...
 ``opclasses`` are ignored for databases besides PostgreSQL.
+
+``db_tablespace``
+-----------------
+
+.. attribute:: UniqueConstraint.db_tablespace
+
+.. versionadded:: 4.0
+
+The name of the :doc:`database tablespace </topics/db/tablespaces>` to use for
```
Comment (felixxm) [inline docs/ref/models/constraints.txt:214]: We should update ["Declaring tablespaces for indexes"](https://docs.djangoproject.com/en/3.2/topics/db/tablespaces/#declaring-tablespaces-for-indexes) to the _"Declaring tablespaces for indexes and unique constraints"_.

File: tests/schema/tests.py [hunk 5]
```diff
@@ -2454,6 +2454,85 @@ def test_func_unique_constraint_nondeterministic(self):
...
+        self.assertIs(sql.references_column(table, 'name'), True)
+        self.assertIn(
+            'WHERE %s IS NOT NULL' % editor.quote_name('weight'),
+            str(sql),
+        )
+        self.assertIn('TABLESPACE %s' % editor.quote_name(constraint.db_tablespace), str(sql))
+        # Remove constraint.
+        with connection.schema_editor() as editor:
+            editor.remove_constraint(Author, constraint)
+        self.assertNotIn(constraint.name, self.get_constraints(table))
```
Comment (felixxm) [inline tests/schema/tests.py:2523]: These tests will not work on Oracle, we would need to create a custom tablespace for them. I think it's enough to check that a properly quoted tablespace name is in the generated SQL like we do for indexes, see 
https://github.com/django/django/blob/e513fb0e77baf2ebcbf2cbe366bdf0228d01119f/tests/model_indexes/tests.py#L286-L320

File: django/db/backends/base/schema.py (hunk 1 above)
Comment (felixxm) [reply django/db/backends/base/schema.py:1261]: Yeah it's quite inconsistent :disappointed:. I think we should change `columns` to `fields` in a separate PR.

**Review by felixxm — COMMENTED** [review]
@hannseman Thanks for this patch :+1:

File: django/db/backends/base/schema.py (hunk 1 above)
Comment (hannseman) [reply django/db/backends/base/schema.py:1262]: Created https://github.com/django/django/pull/14459

File: docs/ref/models/constraints.txt (hunk 4 above)
Comment (hannseman) [reply docs/ref/models/constraints.txt:214]: I'll update `tablespace.txt` after https://github.com/django/django/pull/14459 is merged.

File: tests/schema/tests.py (hunk 5 above)
Comment (hannseman) [reply tests/schema/tests.py]: Thanks! I guess we can still use the PostgreSQL test `test_opclass_tablespace` to still test the actual creation with the `pg_default` tablespace.

File: django/db/backends/base/schema.py (hunk 3 above)
Comment (hannseman) [reply django/db/backends/base/schema.py]: I made `_get_index_tablespace_sql` return an empty string if the backend does not support custom tablespaces. Should make index parts using it behave the same as well.

**Comment by hannseman** [comment]
I can for some reason not reproduce the test failures locally but my guess is that they stem from `_delete_unique_sql` not being aware of tablespaces on fields and thus trying to use `sql_delete_unique` instead of `sql_delete_index`. One option here could be to pass the affected fields to `_delete_unique_sql` as a new `fields` argument and then use `_get_index_tablespace_sql` to check that SQL statement should be used. 

But I wonder if we really should go with this route as it would be a backwards compatible change if I'm not missing anything, that is unique constraints on fields which have custom tablespaces created before the introduction of `_get_index_tablespace_sql` in `_create_unique_sql` would be created as regular constraints and if they are removed after this change they will crash due to the delete index syntax being used instead. 

Maybe it could be OK require users to specify the `db_tablespace` on `UniqueConstraint` and not have the implicit usage for single field constraints? This would mean that we pass `fields=[]` to `_get_index_tablespace_sql` in `_create_unique_sql`

**Comment by felixxm** [comment]
> I can for some reason not reproduce the test failures locally but my guess is that they stem from `_delete_unique_sql` not being aware of tablespaces on fields and thus trying to use `sql_delete_unique` instead of `sql_delete_index`. Not really sure why the failing tests are seeing this as the ones I've looked at do not use any custom tablespaces.

Tests crash on CI because it sets `DEFAULT_INDEX_TABLESPACE` for PostgreSQL. Deleting `unique_together` constraints crashes because now they're aware of tablespaces and we create them as unique indexes. Unfortunately `alter_unique_together()` doesn't distinct these two cases and it's trying to delete a constraint instead of index, e.g.

:green_circle: `CREATE UNIQUE INDEX "test_alunto_pony_pink_weight_131b79dd_uniq" ON "test_alunto_pony" ("pink", "weight") TABLESPACE "ram"; args=()`

:red_circle: `ALTER TABLE "test_alunto_pony" DROP CONSTRAINT "test_alunto_pony_pink_weight_131b79dd_uniq"; args=()`

**Comment by hannseman** [comment]
> Tests crash on CI because it sets DEFAULT_INDEX_TABLESPACE for PostgreSQL. 

Thanks, the setting explains it! 

> Deleting unique_together constraints crashes because now they're aware of tablespaces and we create them as unique indexes. Unfortunately alter_unique_together() doesn't distinct these two cases and it's trying to delete a constraint instead of index.

What do you think about making this change, it looks pretty straightforward to pass the `fields` to the relevant delete functions and control the behaviour of deleting constraints vs indexes. But by making this change we'll flip the case that you describe so when a user tries to remove a constraint created in a previous version which in the new version should've gotten a tablespace set either by `db_tablespace` on the field/model or `DEFAULT_INDEX_TABLESPACE`  they'll see an error about an index not existing. 🤔

**Comment by felixxm** [comment]
> What do you think about making this change, it looks pretty straightforward to pass the `fields` to the relevant delete functions and control the behaviour of deleting constraints vs indexes. But by making this change we'll flip the case that you describe so when a user tries to remove a constraint created in a previous version which in the new version should've gotten a tablespace set either by `db_tablespace` on the field/model or `DEFAULT_INDEX_TABLESPACE` they'll see an error about an index not existing. thinking
> 
> The only other option I see is to only allow customizing the tablespace for a unique constraint by explicitly specifying the `db_tablespace` on the constraint.

Not sure, I don't see a clear path to move it forward. You can a discussion on the [mailing list](https://groups.google.com/g/django-developers) or forum, where you'll reach a wider audience and see what other think.

Closing due to inactivity and lack of consensus.
